[API] Design and Implement POST endpoint to post a MetricScore for LEADERBOARDS

fredricngo opened the pull request:
To address issue #699, I've designed and implemented a post endpoint that allows a logged-in user to post a score to an existing leaderboard for a game.

fredricngo committed:
[API] WIP: Added leaderboard post endpoint to
post scores to the leaderboard for a game.

Works locally, but needs to be tested after
PR #777 has been merged. Addresses issue #699.

fredricngo committed:
Merge remote-tracking branch 'origin/dev' into api/leaderboard-post

fredricngo committed:
[api] WIP: Changed retrieving user during leaderboard post

fredricngo committed:
Removed unused testing libraries in views file

fredricngo committed:
Merge branch 'dev' into api/leaderboard-post

fredricngo committed:
Fixed trailing issues

theworldofobi commented:
Great, I see that the CI tests are passing, let me give it a look!

theworldofobi committed:
Merge branch 'dev' into api/leaderboard-post

theworldofobi reviewed (approved):
I saw some sort of merge conflict, it was super minor and dealt with imports at the top of the file. LGTM!

MuyanXie commented:
@chrislowzhengxi @fredricngo can you provide a sample payload for me to post to the sample url so that I can verify?

MuyanXie reviewed (commented):
see above comments

fredricngo commented:
Hi @MuyanXie. In order to test the code, you will have to have the following objects created in your django shell:

A User with UserProfile
A Game
A Leaderboard (linked to the Game object you created)
A Metric (linked to the Game object you created)
A Lobby and Match (linked to the Game object you created)

Once these objects have been created, login to chigame and access [http://127.0.0.1:8000/api/games/{game_id}/scores/](http://127.0.0.1:8000/api/games/%7Bgame_id%7D/scores/). In the "HTML form" for POST:

Score: 100
Metric id: 1
Match id: 1
(assuming you're using the first Metric and Match objects you previously created).

You should receive a similar response:

HTTP 201 Created
Allow: GET, POST, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "id": 8,
    "score": 100,
    "user": 3,
    "metric": 1,
    "match": 1,
    "leaderboard_entry": 2
}

MuyanXie commented:
@chrislowzhengxi why did you make changes to leaderboards' model?

MuyanXie commented:
I assume their team should be the one doing all the changes?

fredricngo committed:
Merge branch 'dev' into api/leaderboard-post

fredricngo committed:
[api] #968, wip: resolved merge conflicts and solved pre-commit formatting

MuyanXie commented:
@fredricngo @chrislowzhengxi Hi I just sent some error log in slack -- basically it says a table is missing

fredricngo committed:
[api] wip: continuing progress on issue #699.

I was facing migration issues with the leaderboard
app since I was using obsolete migrations. I had to
delete the old migrations and create new ones.

The POST endpoint for the leaderboard now works.

fredricngo committed:
Merge branch 'dev' into api/leaderboard-post

fredricngo committed:
[api] wip: fixed linter errors and made the branch
up to date with dev. closes #968

fredricngo commented:
While working on this PR, I had a database issue. The problem was that the 'LeaderboardPrviacySetting' model exists in the codebase (in leaderboards/models.py), however the corresponding database table was missing. Django believed the migrations were applied, but the actual SQL to make the table was never executed. I think this was due to the migrations being created but not being fully applied. 

To fix, I used makemigrations to create the proper migration file and then used `python manage.py migrate leaderboards --fake-initial` to properly synchronize Django's migration state with the database. 

For other developers, if you encounter a "table does not exist" error as well, you may need to do the same! 

Users should be able to make a POST to MetricScore now.

fredricngo committed:
Merge branch 'dev' into api/leaderboard-post

fredricngo committed:
[api] wip: closes #699. fixed linter tests.

fredricngo commented:
Following this, in collaboration with the Leaderboards team, we discovered that an initial migration file for the leaderboard models was not generated since the init file was missing. After addressing the issue, the POST endpoint should now be accessible for testing.

MuyanXie committed:
Merge branch 'dev' into api/leaderboard-post

MuyanXie reviewed (approved):
LGTM! The endpoint is working as expected. In the future maybe we can have more systematic way of managing errors?

MuyanXie commented:
Issue Score: excellent

Comments:
The PR history is bit messy but overall it's documenting everything very clearly. Thanks!

<!-- {"score": "excellent", "minor": false} -->
